feat(scanoss-agent): Initial version of scanoss agent for Fossology

scanoss-qg opened the pull request:
This PR relates to Add SCANOSS Agent #2166
The agent detects licenses by querying file info from osskb DB. This agent fingerprints all the files related to the upload and query information about that file. Only licenses are kept, other information is discarded. The scan results are placed on license_file table.

## Description

This agent fingerprints all the files related to the upload and query information about that file. Only licenses are kept, other information is discarded. The scan results are placed on license_file table.

### Changes

This is the initial commit that only report license information. Other useful information such as copyright and version could be included

## How to test
Just do an upload and check on "SCANOSS Toolkit" from optional analysis

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2167"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

scanoss-qg commented:
We have updated mod_deps file in order to install library dependencies. Please, verify buiding process again.

ag4ums commented:
@scanoss-qg, I just tested this PR, I have few questions
the scanner stamps the files with the license -> that's great,
however user does not have a way to identify/verify that why a particular file is marked with a particular license (may be i am missing something).
Also, I think a snippet scanner, should show the matched snippets found from different files, don't see anything similar here.

scanoss-qg commented:
Dear Anupam,

As far as we understand, FOSSology takes a single field of data from the agent: License name. We can of course provide you with a comprehensive array with matching data (range of lines match, component name, vendor, PURL, download URL, etc,etc). However, providing any of these extra metadata would require an update to FOSSology’s data model.

Having said that, the scope of this first agent is to provide you the ability to detect a license on a file that not necesarily contains a license header. It is understood that, unlike all other agents, the license provided by SCANOSS is obtained by comparing the file (or snippets) against our Open Source Knowledge.

We understand that this initial agent already brings to FOSSology a capability not available until now. Would it make sense to merge this contribution and then discuss in detail any further integration?

It would be nice to hear your feedback

Best Regards,

ag4ums commented:
hello @scanoss-qg, let me discuss this with the community members, as I think this feature in its current from is not mergeable.
* snippet requires a dedicated page, with consolidated information for snippet findings (copyright page/software heritage as an example).
* you can also consider filling the package information in the conf page... as many of the fields are already available, but need to fill manually.

scanoss-qg commented:
Hi again @ag4ums:
We went through a second round with the agent. First of all, we have changed the agent name in order to not be confusing. We have also added the matching file info on the [Info page]. Now, you can find information such as purl, download url, remote matching file and line ranges in case of snippet match, and path within the original project. We are not currently retrieving copyrights on our free service, so we have not included in this agent version. However, we have tested the Idea on an internal agent version (a tab between scancode findings and fossology findings).
About filling automatically the information on conf section, I suggest letting that for the next agent version. That can Undoubdately be done. We will be getting forward to hear your suggestions and comments.

GMishx commented on src/scanoss/agent/external/inc/json.h:
```diff
@@ -0,0 +1,281 @@
+
```
Can we use the [json-c](https://github.com/json-c/json-c) instead? It is already used in [nomos to create JSON output](https://github.com/fossology/fossology/blob/master/src/nomos/agent/json_writer.c)

It will be more cleaner and easier to maintain.

scanoss-qg replied:
Hi @GMishx, This file is part of our original Scanner CLI source code (https://github.com/scanoss/scanner.c). Let me check if we can publish a modified version.

scanoss-qg replied:
Hi @GMishx. This file is part of our original Scanner CLI (https://github.com/scanoss/scanner.c) and it could be updated every time we release a new version of it. Let me check with the team if we can publish a modified version of the Scanner.

scanoss-qg reviewed (commented):
It seems all changes are there

ag4ums commented:
> It seems all changes are there

looking into it....

ag4ums commented:
> Hi @ag4ums , Sorry for the insistence. Did you have the chance to take a look on the new version of the agent? Once merged this basic functionality , we can continue adding more features. Thanks in advance!

Hi @scanoss-qg , I have looked into it... many thanks for this PR,... as of now it's like other agent that find the licenses of a file,.....the snippet finding information are still missing...may be they are for future implementations.... what may be good here is a plan of all the features that you are planning to implement.. may be an issue....with some milestone...  may be you can also consider joining our community meeting to discuss it(link is in wiki).  ...... in current situation I would like to merge this PR as an experimental feature until we have basic snippet scanning feature included .... we can discuss this once you have the issues/feature list created.

GMishx commented on src/lib/php/Data/AgentRef.php:
```diff
@@ -34,7 +34,8 @@ class AgentRef
     'ojo' => 'O',
     'scancode' => 'S',
     'spasht' => 'Sp',
-    'reso' => 'Rs'
+    'reso' => 'Rs',
+    'scanoss' => 'Sc'
```
@ag4ums should the `scancode` be renamed to `Sc` and `scanoss` to `So`?

GMishx replied:
/cc @ag4ums

GMishx commented on src/scanoss/Makefile:
```diff
@@ -0,0 +1,58 @@
+######################################################################
+# Copyright (C) 2011 Hewlett-Packard Development Company, L.P.
```
Please update the copyright information.

scanoss-qg replied:
Solved

GMishx commented on src/scanoss/agent/Makefile:
```diff
@@ -0,0 +1,68 @@
+######################################################################
+# Copyright (C) 2010-2011 Hewlett-Packard Development Company, L.P.
```
Please update the copyright information.

scanoss-qg replied:
Solved

GMishx commented on src/scanoss/agent/external/inc/log.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <stdbool.h>
+
+// Log file for analytics
+#define WAYUU_ANALYTICS_LOG "/var/log/wayuu-access.log"
```
Can this line be removed if no used.

scanoss-qg replied:
Line Removed

GMishx commented on src/scanoss/agent/external/src/log.c:
```diff
@@ -0,0 +1,101 @@
+
```
Can a license header be added to the file?

scanoss-qg replied:
License added

GMishx replied:
Hello @scanoss-qg , unfortunately I can not see any license added to this file on GitHub.

GMishx commented on src/scanoss/agent/main.c:
```diff
@@ -0,0 +1,179 @@
+
+
+/*!
+ * \brief main function for the pkgagent
```
Please change comment.

```suggestion
 * \brief main function for the scanoss
```

GMishx commented on src/scanoss/agent/snippet_scan.h:
```diff
@@ -0,0 +1,49 @@
+/**************************************************************
+ Copyright (C) 2010-2011 Hewlett-Packard Development Company, L.P.
```
Please update the copyright information.

scanoss-qg replied:
Solved

GMishx replied:
I can still see `Copyright (C) 2010-2011 Hewlett-Packard Development Company, L.P.`

Please check @scanoss-qg

GMishx commented on src/scanoss/agent/src/blacklist_ext.h:
```diff
@@ -0,0 +1,145 @@
+char *BLACKLISTED_EXTENSIONS[] = {
```
Please add appropriate license header.

GMishx commented on src/scanoss/agent/src/scanner.c:
```diff
@@ -0,0 +1,950 @@
+
+#define _GNU_SOURCE
+#include <ctype.h>
+#include <openssl/md5.h>
+#include <openssl/ssl.h>
```
Please consider using [gcrypt](https://www.gnupg.org/documentation/manuals/gcrypt/) instead of openssl since openssl is licensed under Apache v2 from release 3.0.

Please check https://github.com/fossology/fossology/pull/1545/files for usage of hashing functions.

scanoss-qg replied:
Done! Openssl and ssl dependencies were removed. Using GCrypt instead

GMishx commented on src/scanoss/mod_deps:
```diff
@@ -0,0 +1,102 @@
+# FOSSology mod_deps script
+# This script helps you install dependencies on a system. for a module
+#
+# Copyright (C) 2011 Hewlett-Packard Development Company, L.P.
```
Please update the copyright information.

GMishx commented on src/scanoss/ui/Makefile:
```diff
@@ -0,0 +1,39 @@
+######################################################################
+# Copyright (C) 2011 Hewlett-Packard Development Company, L.P.
```
Please update the copyright information.

GMishx commented on src/scanoss/ui/agent-scanoss.php:
```diff
@@ -0,0 +1,79 @@
+<?php
+/***********************************************************
+ Copyright (C) 2009-2013 Hewlett-Packard Development Company, L.P.
+ Copyright (C) 2015, Siemens AG
```
Please update the copyright information.

GMishx commented on src/scanoss/ui/template/scanoss.html.twig:
```diff
@@ -0,0 +1,26 @@
+{#
+    SPDX-License-Identifier: GPL-2.0
+    SPDX-FileCopyrightText: 2021 Sarita Singh <saritasingh.0425@gmail.com>  
```
Please update the copyright information.

GMishx commented on src/scanoss/ui/template/scanoss.html.twig:
```diff
@@ -0,0 +1,26 @@
+{% import "include/macros.html.twig" as macro %}
+
+<li>
+  {{ 'ScanCode Toolkit'|trans }}<img src="images/info_16.png" title="{{ 'To scan code and detect licenses, copyrights, author and more.'|trans }}" alt="" class="info-bullet"/>, {{'scan for'|trans }}<br/>
+  <input type="checkbox" class="browse-upload-checkbox view-license-rc-size" name="scancodeFlags[]" value="license" /> {{ 'License'|trans }}<img src="images/info_16.png" title="{{ 'A full comparison between the database of licenses present in ScanCode and the code uploaded by users based approach'|trans }}" alt="" class="info-bullet"/><br />
+  <input type="checkbox" class="browse-upload-checkbox view-license-rc-size" name="scancodeFlags[]" value="copyright" /> {{ 'Copyright'|trans }}<img src="images/info_16.png" title="{{ 'Copyright along with copyright holder/author information'|trans }}" alt="" class="info-bullet"/><br />
+  <input type="checkbox" class="browse-upload-checkbox view-license-rc-size" name="scancodeFlags[]" value="email" /> {{ 'Email'|trans }}<img src="images/info_16.png" title="{{ 'Email(s) present in the scan code file'|trans }}" alt="" class="info-bullet"/><br />
+  <input type="checkbox" class="browse-upload-checkbox view-license-rc-size" name="scancodeFlags[]" value="url" /> {{ 'URL'|trans }} <img src="images/info_16.png" title="{{ 'URL(s) present in the scan code file'|trans }}" alt="" class="info-bullet"/>
```
Please add `data-toggle="tooltip"` attribute to all tool-tip images for bootstrap.

GMishx commented on src/www/ui/ui-view-info.php:
```diff
@@ -326,36 +357,38 @@ function ShowPackageInfo($Upload, $Item, $ShowMenu=0)
     $this->dbManager->freeResult($result);
 
-    /** RPM Package Info **/
+    /**
+  * RPM Package Info
+**/
```
Please fix the alignment.

scanoss-qg reviewed (commented):
Looks good

scanoss-qg reviewed (commented):
Guys, All Suggestions have been accepted and corrected. SSL and OpenSSL dependencies have been removed. Using GCrypt library instead.
We have also added a README.md as a short guide

shaheemazmalmmd commented:
@scanoss-qg Can you please update the commit messages as per [CONTRIBUTING.md](https://github.com/fossology/fossology/blob/master/CONTRIBUTING.md). Also please squash the commits to one.

scanoss-qg commented:
@shaheemazmalmmd @GMishx @ag4ums I think that we have commited all the suggestions. Please, let us know if you need further information/modifications.

scanoss-qg commented on src/scanoss/mod_deps:
```diff
@@ -0,0 +1,103 @@
+  case "$DISTRO" in
+    Debian|Ubuntu)
+      apt-get $YesOpt install \
+        libcurl4 libcurl4-gnutls-dev libssl-dev jq
```
libssl dependency was removed from src and makefile. I will remove the dependency from mod_deps too

GMishx reviewed (approved):
Changes looks good.

GMishx commented:
@scanoss-qg looks like some PHP-CS tests are failing: https://github.com/fossology/fossology/runs/6658109368?check_suite_focus=true#step:4:208

Please let us know if you need help fixing them.

scanoss-qg commented:
Dear @GMishx, We have finished fixing the failing checks. We will be waiting for your comments

shaheemazmalmmd commented:
@scanoss-qg : I have observed that there are some warnings in the code that needs to be fixed can you please look into them ??

```
snippet_scan.c: In function 'scanTempFile':
snippet_scan.c:156:33: warning: passing argument 2 of 'runScan' from incompatible pointer type [-Wincompatible-pointer-types]
  156 |     int licCount = runScan(aux, detectedLicenses,purl,url,matchType,oss_lines,filePath);    /* Scan for licenses on the temp file */
      |                                 ^~~~~~~~~~~~~~~~
      |                                 |
      |                                 char **
In file included from snippet_scan.c:26:
snippet_scan.h:48:41: note: expected 'unsigned char **' but argument is of type 'char **'
   48 | int runScan(char *path, unsigned char **licenses, unsigned char *purl,unsigned char *url, unsigned char *matchType, unsigned char *oss_lines, unsigned char *filePath);
      |                         ~~~~~~~~~~~~~~~~^~~~~~~~
snippet_scan.c:157:15: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  157 |     if(strcmp(matchType,"none") && (!(strcmp(matchType,"file"))||!(strcmp(matchType,"snippet")) )){
      |               ^~~~~~~~~
      |               |
      |               unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:157:46: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  157 |     if(strcmp(matchType,"none") && (!(strcmp(matchType,"file"))||!(strcmp(matchType,"snippet")) )){
      |                                              ^~~~~~~~~
      |                                              |
      |                                              unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:157:75: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  157 |     if(strcmp(matchType,"none") && (!(strcmp(matchType,"file"))||!(strcmp(matchType,"snippet")) )){
      |                                                                           ^~~~~~~~~
      |                                                                           |
      |                                                                           unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:158:121: warning: format '%d' expects argument of type 'int', but argument 3 has type 'long int' [-Wformat=]
  158 |             sprintf(auxSql,"INSERT INTO scanoss_fileinfo (pfile_fk, matchtype, lineranges, purl, url, filepath) VALUES(%d, '%s', '%s', '%s', '%s', '%s');", key,matchType,oss_lines,purl,url,filePath);
      |                                                                                                                        ~^                                   ~~~
      |                                                                                                                         |                                   |
      |                                                                                                                         int                                 long int
      |                                                                                                                        %ld
snippet_scan.c:170:36: warning: pointer targets in passing argument 1 of 'getLicenseId' differ in signedness [-Wpointer-sign]
  170 |           int detLic= getLicenseId(detectedLicense); /* ... from name, get the key of license that matches short_name at license_ref */
      |                                    ^~~~~~~~~~~~~~~
      |                                    |
      |                                    char *
snippet_scan.c:82:33: note: expected 'unsigned char *' but argument is of type 'char *'
   82 | int getLicenseId(unsigned char *name)
      |                  ~~~~~~~~~~~~~~~^~~~
snippet_scan.c:178:43: warning: pointer targets in passing argument 1 of 'getLicenseId' differ in signedness [-Wpointer-sign]
  178 |               int oldLic= getLicenseId(dep->oldName);
      |                                        ~~~^~~~~~~~~
      |                                           |
      |                                           char *
snippet_scan.c:82:33: note: expected 'unsigned char *' but argument is of type 'char *'
   82 | int getLicenseId(unsigned char *name)
      |                  ~~~~~~~~~~~~~~~^~~~
snippet_scan.c:144:12: warning: unused variable 'readSize' [-Wunused-variable]
  144 |     size_t readSize = fread(contents, size, 1, f);
      |            ^~~~~~~~
snippet_scan.c:132:13: warning: variable 'result' set but not used [-Wunused-but-set-variable]
  132 |   PGresult *result;
      |             ^~~~~~
snippet_scan.c: In function 'runScan':
snippet_scan.c:225:15: warning: pointer targets in passing argument 1 of 'sprintf' differ in signedness [-Wpointer-sign]
  225 |       sprintf(apiurl,"-H %s -p %s",ApiUrl,ApiPort);
      |               ^~~~~~
      |               |
      |               unsigned char *
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdlib.h:25,
                 from snippet_scan.h:27,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:34:1: note: expected 'char * restrict' but argument is of type 'unsigned char *'
   34 | __NTH (sprintf (char *__restrict __s, const char *__restrict __fmt, ...))
      | ^~~~~
snippet_scan.c:227:15: warning: pointer targets in passing argument 1 of 'sprintf' differ in signedness [-Wpointer-sign]
  227 |       sprintf(apiurl,"-H %s",ApiUrl);
      |               ^~~~~~
      |               |
      |               unsigned char *
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdlib.h:25,
                 from snippet_scan.h:27,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:34:1: note: expected 'char * restrict' but argument is of type 'unsigned char *'
   34 | __NTH (sprintf (char *__restrict __s, const char *__restrict __fmt, ...))
      | ^~~~~
snippet_scan.c:230:15: warning: pointer targets in passing argument 1 of 'sprintf' differ in signedness [-Wpointer-sign]
  230 |       sprintf(apiurl,"");
      |               ^~~~~~
      |               |
      |               unsigned char *
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdlib.h:25,
                 from snippet_scan.h:27,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:34:1: note: expected 'char * restrict' but argument is of type 'unsigned char *'
   34 | __NTH (sprintf (char *__restrict __s, const char *__restrict __fmt, ...))
      | ^~~~~
snippet_scan.c:230:22: warning: zero-length gnu_printf format string [-Wformat-zero-length]
  230 |       sprintf(apiurl,"");
      |                      ^~
snippet_scan.c:234:13: warning: pointer targets in passing argument 1 of 'sprintf' differ in signedness [-Wpointer-sign]
  234 |     sprintf(key,"-K %s",accToken);
      |             ^~~
      |             |
      |             unsigned char *
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdlib.h:25,
                 from snippet_scan.h:27,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:34:1: note: expected 'char * restrict' but argument is of type 'unsigned char *'
   34 | __NTH (sprintf (char *__restrict __s, const char *__restrict __fmt, ...))
      | ^~~~~
snippet_scan.c:237:13: warning: pointer targets in passing argument 1 of 'sprintf' differ in signedness [-Wpointer-sign]
  237 |     sprintf(key,"");
      |             ^~~
      |             |
      |             unsigned char *
In file included from /usr/include/features.h:461,
                 from /usr/include/x86_64-linux-gnu/bits/libc-header-start.h:33,
                 from /usr/include/stdlib.h:25,
                 from snippet_scan.h:27,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:34:1: note: expected 'char * restrict' but argument is of type 'unsigned char *'
   34 | __NTH (sprintf (char *__restrict __s, const char *__restrict __fmt, ...))
      | ^~~~~
snippet_scan.c:237:17: warning: zero-length gnu_printf format string [-Wformat-zero-length]
  237 |     sprintf(key,"");
      |                 ^~
snippet_scan.c:286:64: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  286 |       if (!fscanf(Fin, "%s", lic) || (lic == NULL) || (!strcmp(lic, "(null)")) || (!strcmp(lic, "null")) || (!strcmp(lic, "")) || (lic[0] == 255))
      |                                                                ^~~
      |                                                                |
      |                                                                unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:286:92: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  286 |       if (!fscanf(Fin, "%s", lic) || (lic == NULL) || (!strcmp(lic, "(null)")) || (!strcmp(lic, "null")) || (!strcmp(lic, "")) || (lic[0] == 255))
      |                                                                                            ^~~
      |                                                                                            |
      |                                                                                            unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:286:118: warning: pointer targets in passing argument 1 of 'strcmp' differ in signedness [-Wpointer-sign]
  286 |       if (!fscanf(Fin, "%s", lic) || (lic == NULL) || (!strcmp(lic, "(null)")) || (!strcmp(lic, "null")) || (!strcmp(lic, "")) || (lic[0] == 255))
      |                                                                                                                      ^~~
      |                                                                                                                      |
      |                                                                                                                      unsigned char *
In file included from snippet_scan.h:29,
                 from snippet_scan.c:26:
/usr/include/string.h:137:32: note: expected 'const char *' but argument is of type 'unsigned char *'
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
snippet_scan.c:288:7: warning: implicit declaration of function 'asprintf'; did you mean 'vsprintf'? [-Wimplicit-function-declaration]
  288 |       asprintf(&licenses[retLicensesCount], "%s", lic);// asprintf allocs memory. Free after use to avoid leaks
      |       ^~~~~~~~
      |       vsprintf
snippet_scan.c:247:8: warning: variable 'dump' set but not used [-Wunused-but-set-variable]
  247 |   char dump;
      |        ^~~~
snippet_scan.c:221:17: warning: unused variable 'port' [-Wunused-variable]
  221 |   unsigned char port[6];
      |                 ^~~~
snippet_scan.c:261:7: warning: ignoring return value of 'fscanf', declared with attribute warn_unused_result [-Wunused-result]
  261 |       fscanf(Fin,"%s %s %s %s %s",matchType,oss_lines,purl,filePath,url);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
snippet_scan.c:234:21: warning: '%s' directive writing up to 63 bytes into a region of size 61 [-Wformat-overflow=]
  234 |     sprintf(key,"-K %s",accToken);
      |                     ^~  ~~~~~~~~
In file included from /usr/include/stdio.h:867,
                 from snippet_scan.h:28,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:36:10: note: '__builtin___sprintf_chk' output between 4 and 67 bytes into a destination of size 64
   36 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   37 |       __bos (__s), __fmt, __va_arg_pack ());
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
snippet_scan.c: In function 'scanTempFile':
snippet_scan.c:158:137: warning: '%s' directive writing up to 4999 bytes into a region of size between 4339 and 4897 [-Wformat-overflow=]
  158 |             sprintf(auxSql,"INSERT INTO scanoss_fileinfo (pfile_fk, matchtype, lineranges, purl, url, filepath) VALUES(%d, '%s', '%s', '%s', '%s', '%s');", key,matchType,oss_lines,purl,url,filePath);
      |                                                                                                                                         ^~                                          ~~~~
In file included from /usr/include/stdio.h:867,
                 from snippet_scan.h:28,
                 from snippet_scan.c:26:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:36:10: note: '__builtin___sprintf_chk' output between 115 and 7670 bytes into a destination of size 5000
   36 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   37 |       __bos (__s), __fmt, __va_arg_pack ());
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.c: In function 'createTables':
main.c:138:1: warning: control reaches end of non-void function [-Wreturn-type]
  138 | }
      | ^
src/scanner.c: In function 'MD5':
src/scanner.c:68:16: warning: unused variable 'l' [-Wunused-variable]
   68 |   unsigned int l = gcry_md_get_algo_dlen(GCRY_MD_MD5); /* get digest length (used later to print the result) */

```

ag4ums commented:
@scanoss-qg , is the scanning working in general? @shaheemazmalmmd , @GMishx 
I am getting this below error, though it shows scanoss agent run successfully, but dont show any findings
from geeky scan details 
![image](https://user-images.githubusercontent.com/6360640/174741767-92ba7cbf-1374-4725-90d1-a3426524f8c6.png)

scanoss-qg commented:
HI @ag4ums! It seems to be a SQL issue caused by a non clean uninstall. I will go back to you soon.

GMishx commented:
When I first run the agent, it sends following line to scheduler:
`2022-07-01 13:26:58 scheduler [31847] :: JOB[-23].scanoss[31918.localhost]: "row number 0 is out of range 0..-1"`

Running the agent, multiple warning and even some segmentation faults were generated:
```
2022-07-01 14:52:09 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "Segmentation fault"
2022-07-01 14:52:09 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "parse error: Unfinished JSON term at EOF at line 2, column 0"
2022-07-01 14:52:09 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "parse error: Unfinished JSON term at EOF at line 2, column 0"
2022-07-01 14:52:09 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "parse error: Objects must consist of key:value pairs at line 5, column 1"
2022-07-01 14:52:09 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "parse error: Objects must consist of key:value pairs at line 5, column 1"
```

Scanner finally failed after scanning 325/2086 files with following message:
```
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "parse error: Unfinished JSON term at EOF at line 2, column 0"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.293: Snippet scan: failed to dump scan results"
2022-07-01 15:01:29 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: "ERROR snippet_scan.c.253: Snippet scan: failed to run scan "
2022-07-01 15:01:30 scanoss [0] :: JOB[4042].scanoss[36976.localhost]: agent was killed by signal: 11.Segmentation fault
```

scanoss-qg commented:
Thanks for your feedback, I really appreciate it.
Can you share me the project you are trying to scan?

Core was generated by `scanner /tmp/scanoss/scanoss.tmp -o
/tmp/scanoss/scanoss.tmp.json -H https://os'.

As far as I can see, there is a problem there with the url of the API.

Could you check the value on Sysconfig table of DB the variablename=ScAPIURL

I will continue woirking today on this, any information is welcome.


On Fri, Jul 1, 2022 at 6:43 AM Gaurav Mishra ***@***.***>
wrote:

> I ran the agent from command line to capture the coredump and found
> following issue:
>
> $ echo "321" | ./src/scanoss/agent/scanoss --userID=3 --groupID=3 --scheduler_start
> VERSION: "4.1.0.17"
>
> OK
> FATAL snippet_scan.c.206: Snippet_scan: The scan throws an invalid result
> Segmentation fault (core dumped)
>
> $ gdb src/scanoss/agent/scanoss /var/crash/core-scanoss-11-1000-1000-41585-1656668103
> #0  0x00007f44bd392217 in fclose () from /lib/x86_64-linux-gnu/libc.so.6
> #1  0x0000562285fb5b53 in scanTempFile (key=16310) at snippet_scan.c:209
> #2  0x0000562285fb5ef2 in ProcessUpload ***@***.***=321) at snippet_scan.c:369
> #3  0x0000562285fb4a4d in main (argc=<optimized out>, argv=<optimized out>) at main.c:248
>
> Here:
> https://github.com/fossology/fossology/pull/2167/files#diff-35a047fd2940234261f524e719c1efcc8e80a90c71e3fdf7dd51d37b1809550cR204-R209
>
> If f == null then fclose(f) will cause issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fossology/fossology/pull/2167#issuecomment-1172154224>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASTI3FF5N32PKV3S4UJN7TTVR24SRANCNFSM5PFAHOIQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

GMishx commented:
> Thanks for your feedback, I really appreciate it.
> Can you share me the project you are trying to scan?

I scanned [logback-v_1.1.7](https://github.com/qos-ch/logback/archive/refs/tags/v_1.1.7.tar.gz).

> Core was generated by `scanner /tmp/scanoss/scanoss.tmp -o /tmp/scanoss/scanoss.tmp.json -H https://os'. As far as I can see, there is a problem there with the url of the API. Could you check the value on Sysconfig table of DB the variablename=ScAPIURL

The URL is correct (`https://osskb.org/scan/direct`). I think gdb trimmed the string to fit in the window.

> I will continue woirking today on this, any information is welcome.

Sure thanks.

GMishx commented:
While you are working on it @scanoss-qg , I would recommend to add following changes as well for Debian packaging:
```diff
diff --git a/debian/control b/debian/control
index 727dd2899..e50afd49b 100644
--- a/debian/control
+++ b/debian/control
@@ -8,7 +8,7 @@ Build-Depends: debhelper (>=9~), libglib2.0-dev, libmagic-dev, libxml2-dev,
  libboost-program-options-dev, libjsoncpp-dev, libjson-c-dev, libpq-dev,
  php7.0-cli|php7.2-cli|php7.3-cli|php7.4-cli, php-mbstring, php-zip,
  php-xml, libboost-system-dev, libboost-filesystem-dev, libgcrypt20-dev,
- composer
+ composer, libcurl4, libcurl4-gnutls-dev
 Standards-Version: 3.9.1
 Homepage: https://fossology.org
 
@@ -392,6 +392,19 @@ Description: architecture to fetch license, copyright information from scancode.
  .
  This package contains the scancode agent programs and their resources.
 
+Package: fossology-scanoss
+Architecture: any
+Depends: fossology-common, fossology-ununpack, fossology-wgetagent,
+ ${misc:Depends}
+Description: architecture to fetch license and snippet information from scanoss.
+ The FOSSology project is a web based framework that allows you to
+ upload software to be picked apart and then analyzed by software agents
+ which produce results that are then browsable via the web interface.
+ Existing agents include license analysis, metadata extraction, and MIME
+ type identification.
+ .
+ This package contains the scanoss agent programs and their resources.
+
 Package: fossology-spasht
 Architecture: any
 Depends: fossology-common, ${shlibs:Depends}, ${misc:Depends}
diff --git a/src/scanoss/mod_deps b/src/scanoss/mod_deps
index c267a79da..e3c3e76c9 100755
--- a/src/scanoss/mod_deps
+++ b/src/scanoss/mod_deps
@@ -2,7 +2,7 @@
 ######################################################################
 # SCANOSS Agent for FOSSLogy
 # Copyright (C) 2018-2022 SCANOSS.COM
-# 
+#
 #  This program is free software: you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation, either version 2 of the License, or
@@ -77,7 +77,7 @@ if [[ $BUILDTIME ]]; then
   case "$DISTRO" in
     Debian|Ubuntu)
       apt-get $YesOpt install \
-        libcurl4 libcurl4-gnutls-dev libssl-dev jq
+        libcurl4 libcurl4-gnutls-dev jq
       ;;
     RedHatEnterprise*|CentOS|Fedora)
       yum $YesOpt install \
@@ -91,7 +91,7 @@ if [[ $RUNTIME ]]; then
   case "$DISTRO" in
     Debian|Ubuntu)
       apt-get $YesOpt install \
-        libcurl4 libcurl4-gnutls-dev libssl-dev jq
+        libcurl4 libcurl4-gnutls-dev jq
       ;;
     RedHatEnterprise*|CentOS|Fedora)
       yum $YesOpt install \
```

scanoss-qg commented:
Sure, I will take your advice. Thank you!

On Fri, Jul 1, 2022 at 7:15 AM Gaurav Mishra ***@***.***>
wrote:

> While you are working on it @scanoss-qg <https://github.com/scanoss-qg> ,
> I would recommend to add following changes as well for Debian packaging:
>
> diff --git a/debian/control b/debian/control
> index 727dd2899..e50afd49b 100644--- a/debian/control+++ b/debian/control@@ -8,7 +8,7 @@ Build-Depends: debhelper (>=9~), libglib2.0-dev, libmagic-dev, libxml2-dev,
>   libboost-program-options-dev, libjsoncpp-dev, libjson-c-dev, libpq-dev,
>   php7.0-cli|php7.2-cli|php7.3-cli|php7.4-cli, php-mbstring, php-zip,
>   php-xml, libboost-system-dev, libboost-filesystem-dev, libgcrypt20-dev,- composer+ composer, libcurl4, libcurl4-gnutls-dev
>  Standards-Version: 3.9.1
>  Homepage: https://fossology.org
>  @@ -392,6 +392,19 @@ Description: architecture to fetch license, copyright information from scancode.
>   .
>   This package contains the scancode agent programs and their resources.
>  +Package: fossology-scanoss+Architecture: any+Depends: fossology-common, fossology-ununpack, fossology-wgetagent,+ ${misc:Depends}+Description: architecture to fetch license and snippet information from scanoss.+ The FOSSology project is a web based framework that allows you to+ upload software to be picked apart and then analyzed by software agents+ which produce results that are then browsable via the web interface.+ Existing agents include license analysis, metadata extraction, and MIME+ type identification.+ .+ This package contains the scanoss agent programs and their resources.+
>  Package: fossology-spasht
>  Architecture: any
>  Depends: fossology-common, ${shlibs:Depends}, ${misc:Depends}diff --git a/src/scanoss/mod_deps b/src/scanoss/mod_deps
> index c267a79da..e3c3e76c9 100755--- a/src/scanoss/mod_deps+++ b/src/scanoss/mod_deps@@ -2,7 +2,7 @@
>  ######################################################################
>  # SCANOSS Agent for FOSSLogy
>  # Copyright (C) 2018-2022 SCANOSS.COM-# +#
>  #  This program is free software: you can redistribute it and/or modify
>  # it under the terms of the GNU General Public License as published by
>  # the Free Software Foundation, either version 2 of the License, or@@ -77,7 +77,7 @@ if [[ $BUILDTIME ]]; then
>    case "$DISTRO" in
>      Debian|Ubuntu)
>        apt-get $YesOpt install \-        libcurl4 libcurl4-gnutls-dev libssl-dev jq+        libcurl4 libcurl4-gnutls-dev jq
>        ;;
>      RedHatEnterprise*|CentOS|Fedora)
>        yum $YesOpt install \@@ -91,7 +91,7 @@ if [[ $RUNTIME ]]; then
>    case "$DISTRO" in
>      Debian|Ubuntu)
>        apt-get $YesOpt install \-        libcurl4 libcurl4-gnutls-dev libssl-dev jq+        libcurl4 libcurl4-gnutls-dev jq
>        ;;
>      RedHatEnterprise*|CentOS|Fedora)
>        yum $YesOpt install \
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fossology/fossology/pull/2167#issuecomment-1172185574>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASTI3FC2K24KPT3FK24QBITVR3AK7ANCNFSM5PFAHOIQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

scanoss-qg commented:
I would like to ask you a question that could simplify the agent.
Currently, the agent is wrapping the deprecated version of our scanner
written in C and not contain advanced features. Would it be possible to
replace it with the python client that is currently in production?
I mean, just add the dependency during installation (pip3 install scanoss)
and then the agent can use it directly. Or should we include and compile
also the scanner source code too?
This fact would let us simplify the code, keep the scanner updated and make
scans faster. We are ready to deliver a High Precision Snippet Matching
feature that brings accuracy and precision to the scan.
I'll really apreciate your feedback!


On Fri, Jul 1, 2022 at 7:18 AM Quique Goni ***@***.***> wrote:

> Sure, I will take your advice. Thank you!
>
> On Fri, Jul 1, 2022 at 7:15 AM Gaurav Mishra ***@***.***>
> wrote:
>
>> While you are working on it @scanoss-qg <https://github.com/scanoss-qg>
>> , I would recommend to add following changes as well for Debian packaging:
>>
>> diff --git a/debian/control b/debian/control
>> index 727dd2899..e50afd49b 100644--- a/debian/control+++ b/debian/control@@ -8,7 +8,7 @@ Build-Depends: debhelper (>=9~), libglib2.0-dev, libmagic-dev, libxml2-dev,
>>   libboost-program-options-dev, libjsoncpp-dev, libjson-c-dev, libpq-dev,
>>   php7.0-cli|php7.2-cli|php7.3-cli|php7.4-cli, php-mbstring, php-zip,
>>   php-xml, libboost-system-dev, libboost-filesystem-dev, libgcrypt20-dev,- composer+ composer, libcurl4, libcurl4-gnutls-dev
>>  Standards-Version: 3.9.1
>>  Homepage: https://fossology.org
>>  @@ -392,6 +392,19 @@ Description: architecture to fetch license, copyright information from scancode.
>>   .
>>   This package contains the scancode agent programs and their resources.
>>  +Package: fossology-scanoss+Architecture: any+Depends: fossology-common, fossology-ununpack, fossology-wgetagent,+ ${misc:Depends}+Description: architecture to fetch license and snippet information from scanoss.+ The FOSSology project is a web based framework that allows you to+ upload software to be picked apart and then analyzed by software agents+ which produce results that are then browsable via the web interface.+ Existing agents include license analysis, metadata extraction, and MIME+ type identification.+ .+ This package contains the scanoss agent programs and their resources.+
>>  Package: fossology-spasht
>>  Architecture: any
>>  Depends: fossology-common, ${shlibs:Depends}, ${misc:Depends}diff --git a/src/scanoss/mod_deps b/src/scanoss/mod_deps
>> index c267a79da..e3c3e76c9 100755--- a/src/scanoss/mod_deps+++ b/src/scanoss/mod_deps@@ -2,7 +2,7 @@
>>  ######################################################################
>>  # SCANOSS Agent for FOSSLogy
>>  # Copyright (C) 2018-2022 SCANOSS.COM-# +#
>>  #  This program is free software: you can redistribute it and/or modify
>>  # it under the terms of the GNU General Public License as published by
>>  # the Free Software Foundation, either version 2 of the License, or@@ -77,7 +77,7 @@ if [[ $BUILDTIME ]]; then
>>    case "$DISTRO" in
>>      Debian|Ubuntu)
>>        apt-get $YesOpt install \-        libcurl4 libcurl4-gnutls-dev libssl-dev jq+        libcurl4 libcurl4-gnutls-dev jq
>>        ;;
>>      RedHatEnterprise*|CentOS|Fedora)
>>        yum $YesOpt install \@@ -91,7 +91,7 @@ if [[ $RUNTIME ]]; then
>>    case "$DISTRO" in
>>      Debian|Ubuntu)
>>        apt-get $YesOpt install \-        libcurl4 libcurl4-gnutls-dev libssl-dev jq+        libcurl4 libcurl4-gnutls-dev jq
>>        ;;
>>      RedHatEnterprise*|CentOS|Fedora)
>>        yum $YesOpt install \
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/fossology/fossology/pull/2167#issuecomment-1172185574>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ASTI3FC2K24KPT3FK24QBITVR3AK7ANCNFSM5PFAHOIQ>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

GMishx commented:
The scancode wrapper runs similarly:
https://github.com/fossology/fossology/blob/20ca3a482ce18e85859e92842e94fe06bb226237/src/scancode/agent/scancode_wrapper.cc#L86-L93

And installs the agent as a pip package:
https://github.com/fossology/fossology/blob/20ca3a482ce18e85859e92842e94fe06bb226237/install/fo-install-pythondeps#L163-L176

Just a heads up with the problems we are facing with scancode:
1. The agent calls scancode cli once per file.
2. scancode has a significant bootstrap time.
3. This results in the agent running 3-4 times slower in magnitudes compared to cli.

If scanoss does not have such bootstrap time, it is possible to do it.

What do you think about the suggestion @ag4ums?

ag4ums commented:
completely agree @GMishx, bootstrap time is an issue we see here...  if we can reduce the time somehow for python packages should be fine.

scanoss-qg commented:
Hi @ag4ums and @GMishx : I have fixed the reported issues. It seems that when scanning big projects, some processes were not close. I have fixed that issue. Regarding to adding our official CLI, we should suffer the same issue but we could solve it by recreating the project first and then scan it in one shoot. Let leave this improvement for the next agent release.

scanoss-qg commented:
Thanks @GMishx for your feedback! By default, we are omitting snippet scanning on md files. However, the segmentation fault should not happen, and I think it is related to the json parsing. I will try to reproduce the error. Are you trying to scan logback-v_1.1.7 project? I will make a fix ASAP. Stay tuned! :)

scanoss-qg commented:
Hi @GMishx. It was hard to replicate the error. It seems that some processes opened by the agent were not closed correctly. I have forced the error creating hundreds of dummy processes and then failed. I think the issue is now fixed. I tested with other bigger projects and works fine. I also placed errno on the LOG_ERROR for debugging purposes. However, you consider that some types of non-source files wll be skipped (eg: .md, xls, xml, etc). I will be waitting your feedback!

scanoss-qg commented:
Hi @ag4ums and @GMishx, just a quick comment (and not related directly with the agent). After doing the rebase, we did a fresh installation. The post install script fails because of some error related to the DB connection. We found that the file /usr/local/etc/fossology/Db.conf contains a SPDX header with ";"
Removing those headers lines, solves the problem. If you want, I can create a PR for this.

juliancoccia commented:
Sorry to bump this thread. Is there anything we can do to help you merge this PR?
Meanwhile, for users willing to build the SCANOSS agent, we have brought up a repo with instructions here:
https://github.com/scanoss/fossology

GMishx commented:
Hello @scanoss-qg and @juliancoccia , we were waiting to release FOSSology 4.2.0 before merging this PR. Since we have released 4.2.1, we can proceed with merging the scanoss branch to master.

However, we have recently changed our build system from make to CMake, there are some changes required with the branch.

We have updated the documentation for the same and can be refereed from following wikis:
1. [Install from Source](https://github.com/fossology/fossology/wiki/Install-from-Source)
2. [Building Debian packages with CMake](https://github.com/fossology/fossology/wiki/Building-Debian-packages-with-CMake)

Please feel free to contact us if you need support with CMake.
